File: src/features/unity/bridge/iOS/UnityViewManager.m
//
//  UnityViewManager.m
//  app
//
//  Created by Hong Taeho on 9/23/25.
//  React Native Unity View Manager Registration
//

#import <React/RCTViewManager.h>
#import <React/RCTUIManager.h>

@interface RCT_EXTERN_MODULE(UnityView, RCTViewManager)

// Events
RCT_EXPORT_VIEW_PROPERTY(onUnityReady, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onUnityError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onCharacterStateChanged, RCTDirectEventBlock)

// Commands
RCT_EXTERN_METHOD(sendMessageToUnity:(nonnull NSNumber *)node objectName:(NSString *)objectName methodName:(NSString *)methodName parameter:(NSString *)parameter)
RCT_EXTERN_METHOD(pauseUnity:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(resumeUnity:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(reattachUnityView:(nonnull NSNumber *)node)

@end

File: src/features/unity/bridge/iOS/UnityViewManager.swift
//
//  UnityViewManager.swift
//  app
//
//  Created by Hong Taeho on 9/23/25.
//  React Native Unity View Manager
//

import React
import UIKit

@objc(UnityView)
class UnityViewManager: RCTViewManager {

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    override var methodQueue: DispatchQueue? {
        return DispatchQueue.main
    }

    override func view() -> UIView {
        let unityView = UnityView()
        return unityView
    }

    // MARK: - React Native Props
    // 이벤트들은 .m 파일에서 RCT_EXPORT_VIEW_PROPERTY로 등록됩니다.

    // MARK: - React Native Commands

    // Unity에 메시지 전송
    @objc func sendMessageToUnity(_ node: NSNumber, objectName: NSString, methodName: NSString, parameter: NSString) {
        DispatchQueue.main.async {
            guard let unityView = self.bridge.uiManager.view(forReactTag: node) as? UnityView else {
                print("[UnityViewManager] Could not find UnityView for tag: \(node)")
                return
            }

            unityView.sendMessageToUnity(objectName as String, methodName: methodName as String, parameter: parameter as String)
        }
    }

    // Unity 일시정지
    @objc func pauseUnity(_ node: NSNumber) {
        DispatchQueue.main.async {
            guard let unityView = self.bridge.uiManager.view(forReactTag: node) as? UnityView else {
                return
            }

            unityView.pauseUnity()
        }
    }

    // Unity 재개
    @objc func resumeUnity(_ node: NSNumber) {
        DispatchQueue.main.async {
            guard let unityView = self.bridge.uiManager.view(forReactTag: node) as? UnityView else {
                return
            }

            unityView.resumeUnity()
        }
    }

    // Unity View 재연결
    @objc func reattachUnityView(_ node: NSNumber) {
        DispatchQueue.main.async {
            guard let unityView = self.bridge.uiManager.view(forReactTag: node) as? UnityView else {
                print("[UnityViewManager] Could not find UnityView for tag: \(node)")
                return
            }

            unityView.reattachUnityView()
        }
    }
}
